lib/phase/merge.js

```diff
@@ -0,0 +1,113 @@
+'use strict';
+
+const Hoek = require('@hapi/hoek');
+const TEMPLATE_NAME_REGEX_WITH_NAMESPACE =
+    require('screwdriver-data-schema').config.regex.FULL_TEMPLATE_NAME_WITH_NAMESPACE;
+
+/**
+ * Fetch pipeline template
+ * @method fetchPipelineTemplate
+ * @method
+ * @param   {String}                          pipelineTemplateName                        Pipeline template name
+ * @param   {PipelineTemplateVersionFactory}  config.pipelineTemplateVersionFactory       PiplineTemplateVersion Factory to get pipeline templates
+ * @param   {PipelineTemplateTagFactory}      config.pipelineTemplateTagFactory           PipelineTemplateTag Factory to get pipeline templates tag
+ * @param   {PipelineTemplateFactory}         config.pipelineTemplateFactory              PipelineTemplate Factory to get pipeline templates
+ * @return  {Object}                                                                      Pipeline Template
+ */
+async function fetchPipelineTemplate(
+    pipelineTemplateName,
+    pipelineTemplateVersionFactory,
+    pipelineTemplateTagFactory,
+    pipelineTemplateFactory
+) {
+    const [, namespace, name, versionOrTag] = TEMPLATE_NAME_REGEX_WITH_NAMESPACE.exec(pipelineTemplateName);
+    const templateTag = await pipelineTemplateTagFactory.get({
+        name,
+        namespace,
+        tag: versionOrTag
+    });
+    const version = templateTag ? templateTag.version : versionOrTag;
+    const pipelineTemplate = await pipelineTemplateVersionFactory.getWithMetadata(
+        {
+            name,
+            namespace,
+            version
+        },
+        pipelineTemplateFactory
+    );
+
+    return pipelineTemplate;
+}
+
+/**
+ * Merge parsed yaml setting with pipeline template setting
+ * Pipeline yaml will take precedence over pipeline template
+ * @method mergeSetting
+ * @method
+ * @param   {Object}        parsedDoc              Parsed screwdriver yaml
+ * @param   {Object}        pipelineTemplate       Pipeline template
+ * @return  {Object}                               Merged yaml settings
+ */
+function mergeSetting(parsedDoc, pipelineTemplate) {
```

Comment:
```suggestion
function mergeSetting(parsedDoc, pipelineTemplate) {
    const newPipeline = { ...pipelineTemplate.config };
    const { shared: yamlSharedConfig } = parsedDoc;
    const { shared: templateSharedConfig } = newPipeline;

    const mergedEnvironment = {
        SD_PIPELINE_TEMPLATE_FULLNAME: parsedDoc.template,
        SD_PIPELINE_TEMPLATE_NAME: pipelineTemplate.name,
        SD_PIPELINE_TEMPLATE_NAMESPACE: pipelineTemplate.namespace,
        SD_PIPELINE_TEMPLATE_VERSION: pipelineTemplate.version,
        ...templateSharedConfig.environment,
        ...yamlSharedConfig.environment
    };

    const mergedSettings = {
        ...templateSharedConfig.settings,
        ...yamlSharedConfig.settings
    };

    const mergedImage = yamlSharedConfig.image || templateSharedConfig.image;

    newPipeline.shared = {
        ...templateSharedConfig,
        environment: mergedEnvironment,
        settings: mergedSettings,
        image: mergedImage
    };

    return newPipeline;
}
```

Reply:
` const { shared: yamlSharedConfig } = parsedDoc;` has to stay as `    const yamlSharedConfig = parsedDoc.shared || {};`. It would throw error when you try to do yamlSharedConfig, unless we want to add `if` statement to guard that